File: src/ios/IOSOCRVision.m
#import "IOSOCRVision.h"
#import <Vision/Vision.h>
#import <UIKit/UIKit.h>

@implementation IOSOCRVision

- (void)recognizeText:(CDVInvokedUrlCommand *)command
{
    // 1. Get Base64 string
    NSString* base64String = [command.arguments objectAtIndex:0];
    if (!base64String) {
        CDVPluginResult* result = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:@"No image data provided."];
        [self.commandDelegate sendPluginResult:result callbackId:command.callbackId];
        return;
    }
    
    // 2. Convert Base64 to UIImage
    NSData* imageData = [[NSData alloc] initWithBase64EncodedString:base64String options:0];
    UIImage* uiImage = [UIImage imageWithData:imageData];
    if (!uiImage) {
        CDVPluginResult* result = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:@"Invalid image data."];
        [self.commandDelegate sendPluginResult:result callbackId:command.callbackId];
        return;
    }
    
    // 3. Check for cgImage
    CGImageRef cgImage = uiImage.CGImage;
    if (!cgImage) {
        CDVPluginResult* result = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:@"Failed to retrieve CGImage from UIImage."];
        [self.commandDelegate sendPluginResult:result callbackId:command.callbackId];
        return;
    }
    
    // 4. Create a VNRecognizeTextRequest
    VNRecognizeTextRequest* textRequest = [[VNRecognizeTextRequest alloc] initWithCompletionHandler:^(VNRequest * _Nonnull request, NSError * _Nullable error) {
        if (error) {
            CDVPluginResult* result = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:error.localizedDescription];
            [self.commandDelegate sendPluginResult:result callbackId:command.callbackId];
            return;
        }
        
        NSMutableString *recognizedText = [NSMutableString string];
        for (VNRecognizedTextObservation* observation in request.results) {
            if (![observation isKindOfClass:[VNRecognizedTextObservation class]]) { continue; }
            VNRecognizedText* topCandidate = [[observation topCandidates:1] firstObject];
            if (topCandidate) {
                [recognizedText appendString:topCandidate.string];
                [recognizedText appendString:@"\n"];
            }
        }
        
        CDVPluginResult* successResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsString:recognizedText];
        [self.commandDelegate sendPluginResult:successResult callbackId:command.callbackId];
    }];
    textRequest.recognitionLevel = VNRequestTextRecognitionLevelAccurate;
    
    // (Optionally set languages if iOS >= 13)
    if (@available(iOS 13.0, *)) {
        textRequest.recognitionLanguages = @[@"en-US"];
    }
    
    // 5. Perform the request
    VNImageRequestHandler* requestHandler = [[VNImageRequestHandler alloc] initWithCGImage:cgImage options:@{}];
    dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_HIGH, 0), ^{
        NSError *requestError = nil;
        [requestHandler performRequests:@[textRequest] error:&requestError];
        if (requestError) {
            CDVPluginResult* result = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:requestError.localizedDescription];
            [self.commandDelegate sendPluginResult:result callbackId:command.callbackId];
        }
    });
}

@end

File: src/ios/IOSOCRVision.swift
import UIKit
import Vision
import Cordova

@objc(IOSOCRVision) class IOSOCRVision: CDVPlugin {
    @objc(recognizeText:)
    func recognizeText(command: CDVInvokedUrlCommand) {
        // Get the Base64 image string from the Cordova call
        guard let base64Image = command.argument(at: 0) as? String else {
            let pluginResult = CDVPluginResult(status: .error, messageAs: "No image provided.")
            self.commandDelegate.send(pluginResult, callbackId: command.callbackId)
            return
        }
        
        // Decode Base64 to Data and then to UIImage
        guard let imageData = Data(base64Encoded: base64Image),
              let uiImage = UIImage(data: imageData) else {
            let pluginResult = CDVPluginResult(status: .error, messageAs: "Invalid image data.")
            self.commandDelegate.send(pluginResult, callbackId: command.callbackId)
            return
        }
        
        // Ensure the UIImage can provide a CGImage
        guard let cgImage = uiImage.cgImage else {
            let pluginResult = CDVPluginResult(status: .error, messageAs: "Failed to get CGImage from UIImage.")
            self.commandDelegate.send(pluginResult, callbackId: command.callbackId)
            return
        }
        
        // Create a Vision request handler
        let requestHandler = VNImageRequestHandler(cgImage: cgImage, options: [:])
        let request = VNRecognizeTextRequest { (request, error) in
            if let error = error {
                let pluginResult = CDVPluginResult(status: .error, messageAs: error.localizedDescription)
                self.commandDelegate.send(pluginResult, callbackId: command.callbackId)
                return
            }
            
            // Process the results from the Vision request
            var recognizedText = ""
            for observation in request.results as? [VNRecognizedTextObservation] ?? [] {
                guard let candidate = observation.topCandidates(1).first else { continue }
                recognizedText += candidate.string + "\n"
            }
            
            // Return the recognized text to the Cordova layer
            let pluginResult = CDVPluginResult(status: .ok, messageAs: recognizedText)
            self.commandDelegate.send(pluginResult, callbackId: command.callbackId)
        }
        
        // Configure the request as needed
        request.recognitionLevel = .accurate
        if #available(iOS 13.0, *) {
            request.recognitionLanguages = ["en-US"]
        }
        
        // Perform the Vision request asynchronously
        DispatchQueue.global(qos: .userInitiated).async {
            do {
                try requestHandler.perform([request])
            } catch let error {
                let pluginResult = CDVPluginResult(status: .error, messageAs: error.localizedDescription)
                self.commandDelegate.send(pluginResult, callbackId: command.callbackId)
            }
        }
    }
}
